contracts/external/traders/DaiMigrator.sol

```diff
@@ -0,0 +1,166 @@
+
+    uint256 constant DAI_MARKET = 3;
+
+    // ============ Events ============
+
+    event LogMigratorAdded(
+        address migrator
+    );
+
+    event LogMigratorRemoved(
+        address migrator
+    );
+
+    // ============ Storage ============
+
+    // the addresses that are able to migrate positions
+    mapping (address => bool) public g_migrators;
+
+    // ============ Constructor ============
+
+    constructor (
+        address[] memory migrators
+    )
+        public
+    {
+        for (uint256 i = 0; i < migrators.length; i++) {
+            g_migrators[migrators[i]] = true;
+        }
+    }
+
+    // ============ Admin Functions ============
+
+    function addMigrator(
+        address migrator
+    )
+        external
+        onlyOwner
+    {
+        emit LogMigratorAdded(migrator);
+        g_migrators[migrator] = true;
+    }
+
+    function removeMigrator(
+        address migrator
+    )
+        external
+        onlyOwner
+    {
+        emit LogMigratorRemoved(migrator);
+        g_migrators[migrator] = false;
+    }
+
+    // ============ Only-Solo Functions ============
+
+    function getTradeCost(
+        uint256 inputMarketId,
+        uint256 outputMarketId,
+        Account.Info memory /* makerAccount */,
+        Account.Info memory takerAccount,
+        Types.Par memory oldInputPar,
+        Types.Par memory newInputPar,
+        Types.Wei memory inputWei,
+        bytes memory /* data */
+    )
+        public
+        returns (Types.AssetAmount memory)
+    {
+        Require.that(
+            g_migrators[takerAccount.owner],
+            FILE,
+            "Migrator not approved",
+            takerAccount.owner
+        );
+
+        Require.that(
+            inputMarketId == SAI_MARKET && outputMarketId == DAI_MARKET,
+            FILE,
+            "Invalid markets"
+        );
+
+        // require that SAI amount is getting smaller (closer to zero)
+        if (oldInputPar.isPositive()) {
+            Require.that(
+                inputWei.isNegative(),
+                FILE,
+                "inputWei must be negative"
+            );
+            Require.that(
+                !newInputPar.isNegative(),
+                FILE,
+                "newInputPar cannot be negative"
+            );
+        } else if (oldInputPar.isNegative()) {
+            Require.that(
+                inputWei.isPositive(),
+                FILE,
+                "inputWei must be positive"
+            );
+            Require.that(
+                !newInputPar.isPositive(),
+                FILE,
+                "newInputPar cannot be positive"
+            );
+        } else {
+            Require.that(
+                inputWei.isZero() && newInputPar.isZero(),
+                FILE,
+                "inputWei must be zero"
+            );
+        }
+
+        /* return the exact opposite amount of SAI in DAI */
+        return Types.AssetAmount ({
+            sign: !inputWei.sign,
```

Comment:
Ok so this is acting on the maker account, so it's negative as in it's giving away DAI. Makes sense 👍